File: examples/companion-packages/java-maven-lib/src/main/java/ai/agentxm/examples/tinyflags/EvaluationContext.java
package ai.agentxm.examples.tinyflags;

import java.util.Optional;

/**
 * Identity attributes used to bucket a flag evaluation deterministically.
 *
 * <p>Pass a stable identifier ({@code userId}, {@code accountId}, or
 * {@code sessionId}) so a given caller receives stable rollout decisions.
 * Callers without any identifier share a single "anonymous" bucket.
 */
public record EvaluationContext(
        Optional<String> userId,
        Optional<String> accountId,
        Optional<String> sessionId) {

    /** An evaluation context with no identity attributes. */
    public static final EvaluationContext EMPTY = new EvaluationContext(
            Optional.empty(), Optional.empty(), Optional.empty());

    /** Returns a context with the given user id. */
    public static EvaluationContext ofUser(String userId) {
        return new EvaluationContext(Optional.of(userId), Optional.empty(), Optional.empty());
    }

    /** Returns a context with the given account id. */
    public static EvaluationContext ofAccount(String accountId) {
        return new EvaluationContext(Optional.empty(), Optional.of(accountId), Optional.empty());
    }

    /** Returns a context with the given session id. */
    public static EvaluationContext ofSession(String sessionId) {
        return new EvaluationContext(Optional.empty(), Optional.empty(), Optional.of(sessionId));
    }
}

File: examples/companion-packages/composer-lib/src/EvaluationContext.php
<?php

declare(strict_types=1);

namespace AgentXM\Examples\TinyFlags;

/**
 * Stable identity used for deterministic rollout bucketing.
 *
 * Pass at least one of userId, accountId, or sessionId to receive
 * per-caller bucketing. Callers without any identifier share a single
 * "anonymous" bucket.
 */
final class EvaluationContext
{
    public function __construct(
        public readonly ?string $userId = null,
        public readonly ?string $accountId = null,
        public readonly ?string $sessionId = null,
    ) {
    }

    public function bucketKey(): string
    {
        return $this->userId ?? $this->accountId ?? $this->sessionId ?? 'anonymous';
    }
}
